Use a modern approach for cross-site request forgery protection

rosa opened the pull request:
### Motivation / Background

CSRF protection has always been a source of issues related to cached pages, stale tokens, the session being somehow cleared, all resulting in false positives. We still get these in our apps. 

This PR proposes to modernise this, with the goal of eventually deprecating and getting rid of CSRF tokens completely. It takes advantage of the `Sec-Fetch-Site` header, which is a fetch metadata request header that modern browsers send, and that indicates whether a request for a resource is coming from the same origin, the same site or a different site. The server can use this to decide whether the request should be allowed or not.

In this PR, this approach is used by default, in a very conservative and strict way: only requests with a `Sec-Fetch-Site` value of `same-site` or `same-origin` are allowed. Requests with a value of `cross-site` are rejected unless they come from a configured trusted origin (see below, in the Detail section). Requests that don't include a `Sec-Fetch-Site` header or that have it set to `none` are rejected as well. Assuming a modern browser, user-initiated requests (like direct navigation or clicking a bookmark) shouldn't cause a non-GET/HEAD request, so these wouldn't be subject to the CSRF check in any case. Requests that don't include the header could be:
- Not from a browser, in which case they wouldn't include the authenticity token either. These would be excluded from the check in another way.
- From a browser, but having the header stripped by some middle proxy or firewall (we've seen this in the wild). In this case, we can't allow the request without making the user vulnerable to CSRF.

This is what we're currently using in HEY, with a custom strategy class for unverified request handling. 

For apps that can't use this approach yet because they need to support older browsers that don't send the `Sec-Fetch-Site` header, the usual authenticity token approach is kept as a fallback. When a decision can't be clearly made with the header value because it's missing or 'none', the token is checked as before.

The `Origin` check is not changed; it continues to work in the same way, combined with the `Sec-Fetch-Site `or authenticity token checks.

### Detail
This PR introduces a new option, `using`, that can be passed to `protect_from_forgery`, to indicate the approach used for verifying requests. The possible values are `header_only`, which relies exclusively on the `Sec-Fetch-Site` header, or `header_or_legacy_token`, which uses the authenticity token when a decision can't be made with the header alone. The default would be `header_only` for Rails 8.2. Example: 

```ruby
class ApplicationController < ActionController::Base
  protect_from_forgery using: :header_or_legacy_token, with: :exception
```

When using the `header_or_legacy_token` approach, a warning is logged when the fallback is necessary. In this way, developers can check which requests don't include valid `Sec-Fetch-Site` headers and progressively fix them so that they can eventually stop using the CSRF token. 

A new option, `trusted_origins`, is also added to set trusted origins for legitimate cross-site requests (OAuth callbacks, third-party embeds):
 
```ruby
protect_from_forgery trusted_origins: %w[ https://accounts.google.com ]
```

I'll include some comments in the PR to point out other important details. 

### Additional information
Some references about this: 
- https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/Sec-Fetch-Site
- https://web.dev/articles/fetch-metadata
- https://github.com/golang/go/issues/73626
- https://www.alexedwards.net/blog/preventing-csrf-in-go

### Open questions
Besides the obvious question of whether this is an acceptable approach for Rails and whether this makes sense with the current implementation, a possible open question is: should the values allowed for `Sec-Fetch-Site` be configurable? This would allow for an even stricter policy where `same-site` requests aren't allowed, only `same-origin`. This would prevent requests coming from a different subdomain, for example. It would also allow for a more relaxed policy where `none` or missing values are allowed because the risk of people running older browsers or some kind of proxy is deemed acceptable. 


### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Unrelated changes should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

rosa committed:
Use a modern approach for cross-site request forgery protection

Take advantage of the Sec-Fetch-Site header, which is a fetch metadata
request header that modern browsers send, and that indicates whether a
request for a resource is coming from the same origin, the same site or
a different site. The server can use this to decide whether the request
should be allowed or not.

This approach is used by default, in a very conservative and strict way:
only requests with a Sec-Fetch-Site value of same-site or same-origin
are allowed. Requests that don't include a Sec-Fetch-Site header or that
have it set to none are rejected. Assuming a modern browser, user-initiated
requests (like direct navigation or clicking a bookmark) shouldn't cause
a non-GET/HEAD request, so these wouldn't be subject to the CSRF check
in any case. Requests that don't include the header could either be:
- Not from a browser, in which case they wouldn't include the
  authenticity token either. These would be excluded from the check in
  another way.
- From a browser, but having the header stripped by some middle proxy or
  firewal (we've seen this in the wild). In this case we can't allow the
  request without making the user vulnerable to CSRF.

For apps that can't use this approach yet because they need to support
older browsers that don't send the Sec-Fetch-Site header, the usual
authenticity token approach is kept as a fallback. When a decision can't
be clearly made with the header value because it's missing or 'none',
the token is checked as before.

The Origin check is not changed, it continues to work in the same way,
combined with the Sec-Fetch-Site or authenticity token checks.

References:
- https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/Sec-Fetch-Site
- https://web.dev/articles/fetch-metadata
- https://github.com/golang/go/issues/73626
- https://www.alexedwards.net/blog/preventing-csrf-in-go

jeremy committed:
Add trusted_origins option for cross-site request allowlisting

Adds support for explicitly allowing cross-site requests from trusted
origins, useful for OAuth/SSO callbacks, third-party embeds, and
legitimate remote form submission.

Usage:

    protect_from_forgery trusted_origins: %w[ https://accounts.google.com ]

This works with both :fetch_metadata and :token_fallback verification
strategies. When a cross-site request arrives with an Origin header
matching a trusted origin, it's allowed through without requiring a
CSRF token.

jeremy committed:
Deprecate InvalidAuthenticityToken in favor of InvalidCrossOriginRequest

Since CSRF protection now relies on Sec-Fetch-Site headers rather than
authenticity tokens, InvalidAuthenticityToken no longer accurately
describes the protection mechanism.

Deprecate InvalidAuthenticityToken with a message directing users to
catch InvalidCrossOriginRequest instead. The deprecated constant will
be removed in Rails 9.0.

jeremy commented on actionpack/lib/action_controller/metal/request_forgery_protection.rb:
```diff
@@ -40,7 +46,7 @@ class InvalidCrossOriginRequest < ActionControllerError # :nodoc:
   # Subclasses of ActionController::Base are protected by default with the
   # `:exception` strategy, which raises an
-  # ActionController::InvalidAuthenticityToken error on unverified requests.
+  # ActionController::InvalidCrossOriginRequest error on unverified requests.
```
Can deprecate the constant! https://github.com/rosa/rails/pull/2

jeremy commented on actionpack/lib/action_controller/metal/request_forgery_protection.rb:
```diff
@@ -460,16 +550,63 @@ def non_xhr_javascript_response? # :doc:
+
         request.get? || request.head? || !protect_against_forgery? ||
-          (valid_request_origin? && any_authenticity_token_valid?)
+          (valid_request_origin? && verified_request_for_forgery_protection?)
```
```suggestion
        valid_request_origin? && verified_request_for_forgery_protection?
```

jeremy committed:
Default to token_fallback for CSRF, enable fetch_metadata via load_defaults 8.2

This ensures existing applications upgrading to Rails 8.2 continue using
token-based CSRF protection (with Sec-Fetch-Site optimization) rather than
silently switching to header-only verification which would break older
browsers.

New applications using load_defaults(8.2) get the modern fetch_metadata
strategy that relies solely on Sec-Fetch-Site headers.

jeremy committed:
Rename CSRF verification strategies for clarity

Rename :fetch_metadata to :header_only
Rename :token_fallback to :header_or_legacy_token

The new names are intention-revealing and self-explanatory:

- :header_only - verify requests using Sec-Fetch-Site header exclusively
- :header_or_legacy_token - use header, fall back to authenticity token

"fetch_metadata" was jargon referring to the Fetch Metadata spec that most
Rails developers wouldn't recognize. "token_fallback" didn't explain what
it was falling back from.

The "legacy" modifier in :header_or_legacy_token signals that token-based
CSRF is the old approach and may eventually move to a plugin.

jeremy commented on actionpack/lib/action_controller/metal/request_forgery_protection.rb:
```diff
@@ -106,6 +112,20 @@ module RequestForgeryProtection
 
+      # The strategy to use for verifying requests. Options are:
+      # * :fetch_metadata - Use Sec-Fetch-Site header only (default, modern browsers)
+      # * :token_fallback - Combined approach: Sec-Fetch-Site with fallback to token
```
Suggest clarifying naming: https://github.com/rosa/rails/pull/4

bdewater-thatch commented on actionpack/lib/action_controller/metal/request_forgery_protection.rb:
```diff
@@ -460,16 +552,63 @@ def non_xhr_javascript_response? # :doc:
+        when "cross-site"
+          origin_trusted?
+        else # "none" or missing
+          logger.warn "Falling back to CSRF token check for forgery protection" if logger && log_warning_on_csrf_failure
```
IDK if this should be Rubocop rule or documented somewhere, but I think Rails framework components shouldn't log directly and use Active Support Notifications (or the new Event Reporter) instead so that applications can have control over the log output and hook up their own additional behaviour in a clean way.

rosa replied:
Good point! I had used Active Support notifications on my first approach, but changed to this to match the other logging done in this module. Maybe everything should be switched instead of mixing the two ways?

jeremy replied:
Agree re. sticking with existing usage. We can introduce structured events in a separate PR.

jeremy committed:
Document Sec-Fetch-Site CSRF protection

Add CHANGELOG entry describing the new header-based CSRF protection,
verification strategies, trusted_origins option, and exception deprecation.

Add configuring.md documentation for:
- config.action_controller.forgery_protection_verification_strategy
- config.action_controller.forgery_protection_trusted_origins

Add forgery_protection_verification_strategy to 8.2 defaults list.

rosa committed:
Remove redundant, left-over guard

jeremy committed:
Fix protect_from_forgery documentation structure

Nest verification strategies under :using option and clarify that
:trusted_origins is a separate option, not a verification strategy.

chaadow commented on actionpack/lib/action_controller/metal/request_forgery_protection.rb:
```diff
@@ -203,25 +225,65 @@ module ClassMethods
         self.request_forgery_protection_token ||= :authenticity_token
 
         self.csrf_token_storage_strategy = storage_strategy(options[:store] || SessionStore.new)
+        self.forgery_protection_verification_strategy = verification_strategy(options[:using] || :header_only)
```
I believe this should be 
```suggestion
        self.forgery_protection_verification_strategy = verification_strategy(options[:using] || forgery_protection_verification_strategy)
```

For those who still want to keep the old behaviour

chaadow commented on actionpack/lib/action_controller/metal/request_forgery_protection.rb:
```diff
@@ -203,25 +225,65 @@ module ClassMethods
       def skip_forgery_protection(options = {})
-        skip_before_action :verify_authenticity_token, options.reverse_merge(raise: false)
+        skip_before_action :verify_request_for_forgery_protection, options.reverse_merge(raise: false)
       end
```
I think a `rubocop-rails` cop detecting `skip_before_action :verify_authenticity_token` and changing it to `skip_forgery_protection` can be cool as well ( Just throwing some ideas that I've just thought of)

chaadow replied:
here is an example of the `Gitlab` codebase : https://github.com/gitlabhq/gitlabhq/blob/5395a822abdb4af1415a875a4d70c19c3a3d4118/app/controllers/jwt_controller.rb#L6

rosa replied:
I think we should revert the rename, then, and add a deprecation notice 🤔 Then rename in the next version. I'll open a PR with this proposal.

chaadow replied:
I'll be interested to see how you can add a deprecation notice in this case.
 I might be missing something, but since `protect_from_forgery` is the one calling `before_action :verify_authenticity_token`, and 99,99% of applications call `protect_from_forgery` in their controllers in some form, that means all applications will have this deprecation warning :/
